[CMake] Detect properly new linker introduced in Xcode 15

edymtt opened the pull request:
As explained in [1], this linker is functionally equivalent to the classic one (`ld64`)
for build system purposes -- in particular to enable the use of order files to link
`clang`. For this reason, in addition to fixing the detection rename
`LLVM_LINKER_IS_LD64` to `LLVM_LINKER_IS_APPLE` to make the result of
such detection more clear -- this should not cause any issue to downstream
users, from a quick search in SourceGraph [2], only Swift uses the value of
this variable (which I will take care of updating in due time).

[1]: https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes#Linking
[2]: https://sourcegraph.com/search?q=context:global+LLVM_LINKER_IS_LD64+lang:cmake+fork:no+-file:AddLLVM.cmake+-file:clang/tools/driver/CMakeLists.txt&patternType=standard&sm=1&groupBy=repo
rdar://120740222

edymtt commented:
macOS testing triggered on https://github.com/apple/llvm-project/pull/7962

edymtt commented:
(Rebased on top of recent changes to take format changes in `clang/docs/HLSL/FunctionCalls.rst`

ldionne commented on llvm/cmake/modules/AddLLVM.cmake:
```diff
@@ -211,10 +211,10 @@ if (NOT DEFINED LLVM_LINKER_DETECTED AND NOT WIN32)
 
   if(APPLE)
-    if("${stderr}" MATCHES "PROJECT:ld64")
+    if("${stderr}" MATCHES "PROGRAM:ld")
```
I'm not certain I understand -- isn't this patch missing something so that it keeps working on pre Xcode 15 toolchains? Libc++ supports Xcode 15 only, but the rest of LLVM probably supports older versions as well?

edymtt replied:
Yes, that's the intent -- in fact we are not checking the `PROJECT` part anymore, but the `PROGRAM` one, which I found empirically remains more "stable" across Xcode versions (while not making the check needlessly complicated with regexes):

```
> .../Xcode14.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld -v
@(#)PROGRAM:ld  PROJECT:ld64-857.1
...
```

```
> .../Xcode15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld -v
@(#)PROGRAM:ld  PROJECT:dyld-1022.1
...
```

ldionne replied:
Ahhh, now this all makes sense! I hadn't noticed that you switched to `PROGRAM:ld`.

ldionne commented:
The CI failures seem unrelated. I'll merge when the hidden e-mail address issue has been resolved.

edymtt committed:
[CMake] Detect properly new linker introduced in Xcode 15

As explained in [1], this linker is functionally equivalent to the
classic one (`ld64`) for build system purposes -- in particular to
enable the use of order files to link `clang`.
For this reason, in addition to fixing the detection rename
`LLVM_LINKER_IS_LD64` to `LLVM_LINKER_IS_APPLE` to make the result of
such detection more clear -- this should not cause any issue to
downstream users, from a quick serch in
Sourcegraph [2], only Swift uses the value of this variable (which I
will take care of updating in due time).

[1]:
https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes#Linking
[2]:
https://sourcegraph.com/search?q=context:global+LLVM_LINKER_IS_LD64+lang:cmake+fork:no+-file:AddLLVM.cmake+-file:clang/tools/driver/CMakeLists.txt&patternType=standard&sm=1&groupBy=repo
rdar://120740222

edymtt commented:
That issue should now be addressed in https://github.com/llvm/llvm-project/commit/1530d4c137642715d11b5b8b8e2aefcb7566d7c9 (for future reference, this concerned is discussed in [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223), not in the contribution documentation)

Also matched the commit message to your edits (which I forgot to thank you for, the format changes improves the overall readability)
